### PR TITLE
Add Windows GetFolderPath implementation

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Add Environment.GetFolderPath() implementation for Windows.
Using SHGetKnownFolderPath as it isn't blocked at MAX_PATH.
Tweak the .json files to reduce warnings.

This needs an updated build tools- https://github.com/dotnet/buildtools/pull/1022.
I've validated with the relevant build tool change manually.

@weshaggard, @danmosemsft, @ericstj, @danmosemsft 

Also note that I need to make a similar change to NetFX/desktop to allow for long paths.

### CODE REVIEW

File: src/System.Runtime.Extensions/src/System/Environment.Win32.cs [hunk 1]
```diff
@@ -208,14 +208,185 @@ private unsafe static char[] GetEnvironmentCharArray()
...
+            // We're using SHGetKnownFolderPath instead of SHGetFolderPath as SHGetFolderPath is
+            // capped at MAX_PATH.
+            //
+            // Because we validate both of the input enums we shouldn't have to care about CSIDL and flag
+            // definitions we haven't mapped. If we remove or loosen the checks we'd have to account
+            // for mapping here (this includes tweaking as SHGetFolderPath would do).
+            //
+            // The only SpecialFolderOption defines we have are equivalent to KnownFolderFlags.
+
+            string folderGuid = null;
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.Win32.cs]: Nit: can you remove the ```= null```?  Then the compiler should warn if any path in the switch doesn't assign it.

File: src/System.Runtime.Extensions/src/System/Environment.Win32.cs [hunk 2]
```diff
@@ -208,14 +208,185 @@ private unsafe static char[] GetEnvironmentCharArray()
...
+        {
+            Guid folderId = new Guid(folderGuid);
+
+            string path;
+            int hr = Interop.mincore.SHGetKnownFolderPath(folderId, (uint)option, IntPtr.Zero, out path);
+            if (hr != 0) // Not S_OK
+            {
+                if (hr == Interop.mincore.COR_E_PLATFORMNOTSUPPORTED)
+                {
+                    throw new PlatformNotSupportedException();
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Environment.Win32.cs]: Why do we throw here rather than just returning string.Empty?  In what situation will this return PLATFORMNOTSUPPORTED?

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, SetLastError = false, BestFitMapping = false, ExactSpelling = true)]
+        internal static extern int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: If this is marshaling from a PWSTR to a string, don't we need to specify the CharSet as Unicode and/or use UnmanagedType.LPWStr on the parameter?

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+internal partial class Interop
+{
+    internal partial class mincore
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, SetLastError = false, BestFitMapping = false, ExactSpelling = true)]
+        internal static extern int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: I think this is fine (i.e. the one use case for LPStruct), but out of curiosity why did you choose to use such a marshaling directive over just doing ```ref Guid rfid```?

File: src/System.Runtime.Extensions/src/System/Environment.Win32.cs (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.Win32.cs]: Done

File: src/System.Runtime.Extensions/src/System/Environment.Win32.cs (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/src/System/Environment.Win32.cs]: This is to match the desktop implementation. I didn't find a source for the error.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 4 above)
Comment (JeremyKuhne) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: Because it isn't actually passed out- it is clearer on intent.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: Odd, I don't know why it didn't matter in this case. I'll add it to the import.

File: src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json [hunk 5]
```diff
@@ -7,6 +7,7 @@
     "System.Threading.Thread": "4.3.0-beta-devapi-24503-01"
   },
   "frameworks": {
+    ".NETCoreApp,Version=v1.0": {},
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: Why did you do the full name instead of netcoreapp1.0?

Comment (weshaggard) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: Actually why did you need to add these at all? I don't see any new build configurations in this PR that would need them.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+internal partial class Interop
+{
+    internal partial class mincore
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true)]
+        internal static extern int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: @yizhang82, is there a performance advantage one way or the other during marshaling for:
```
[MarshalAs(UnmanagedType.LPStruct)] Guid rfid
```
vs
```
ref Guid rfid
```
?  My understanding is that both are correct behaviorally.

**Comment by danmosemsft** [comment]
:shipit:

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 6 above)
Comment (yizhang82) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: [MarshalAs(LpStruct)] Guid does additional copy internally, while ref GUID passes directly by-ref, so ref Guid is slightly better. Also, [in] ref Guid is the better option than ref Guid since ref Guid is [in, out] ref Guid and implies marshalling in both directions, which isn't the case here. 
In this case, either way it makes no practical difference.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 3 above)
Comment (yizhang82) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: @JeremyKuhne By default, we pick ANSI version, which has correct behavior but is less efficient. Passing CharSet.Unicode is better since it tells interop to pick the unicode version. Specifying UnmanagedType.LPWStr without specifying CharSet.unicode is dangerous because we might end up picking the ANSI version of the API while marshal the string as Unicode, unless you expliclty pick the 'W' version in its name.

Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: Is there an ANSI version?  https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188(v=vs.85).aspx shows SHGetKnownFolderPath taking a LPWSTR, and it doesn't list an SHGetKnownFolderPathA.  Maybe I'm just missing it?

File: src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: The build output told me to. I can remove it if you want.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: There is no ANSI version afaik (no reference to 'A' version in Windows sources)- the shell APIs that were added in Vista and later were all Unicode I believe.

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 6 above)
Comment (JeremyKuhne) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: @yizhang82 Thanks for the details. When set to `ref` it should pass by ref as it is a blittable struct, correct? (It is on the stack and pinned by definition, right?) `[In] ref` shouldn't make any difference because of the use of the direct address? (Other than indicating the intent.)

File: src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs (hunk 3 above)
Comment (yizhang82) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: You are right. I was talking about the general case. In this case it makes no difference.

Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: > In this case it makes no difference.

You would still need the UnmanagedType.LPWStr on the parameter or CharSet.Unicode on the DllImport, though, right?  Otherwise how does the runtime know to use Unicode marshaling rather than ANSI for that parameter?  Wouldn't it default to ANSI and do the wrong thing?

Comment (yizhang82) [inline src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs]: Correct. What I meant was there is no practical difference between using either one of them. But you do need one :)

**Comment by danmosemsft** [comment]
@dotnet-bot test this please (all jobs failed, with broken links to details)

File: src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json (hunk 5 above)
Comment (ericstj) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: That may mean a mis-configured builds file.  /cc @karajas

Comment (ericstj) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: Actually more likely a missing configuration for NuGetTargetMoniker in the CSProj.

File: src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs [hunk 7]
```diff
@@ -246,6 +246,65 @@ public void GetFolderPath_OSX_NonEmptyFolderPaths(Environment.SpecialFolder fold
             }
         }
 
+        [Theory]
```
Comment (ericstj) [inline src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs]: This test needs to be in a separate file that is only included in netstandard1.7 build of test assembly.  That is the cause of your CI failures.

Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/System/EnvironmentTests.cs]: I'll move it- I'm also failing as I need to update build tools.  https://github.com/dotnet/buildtools/pull/1022

File: src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.Runtime.Extensions/tests/VoidMainWithExitCodeApp/project.json]: Ok- I'll look.

**Comment by joperezr** [comment]
@JeremyKuhne same for this one. Could you update/close this PR so that we can close dev/api branch soon?

**Comment by JeremyKuhne** [comment]
@joperezr I'll move this over to master- needs a build tools update

**Comment by joperezr** [comment]
@JeremyKuhne so do you want me to close this one as well?

**Comment by JeremyKuhne** [comment]
@joperezr I'll take care of closing it after I open the new PR

**Comment by JeremyKuhne** [comment]
Moving to master.
